### PR TITLE
👌 IMPROVE: Replace deprecated imp with importlib

### PR DESCRIPTION
**Pull request by DirectriX01**
I've replaced it according to the official  [Importlib documentation](https://github.com/aiidateam/aiida-core/blob/dc9ec6f91e0b23608c58ece6e3283a13780dcf47/aiida/orm/nodes/data/cif.py#L49)

### CODE REVIEW

**Comment by chrisjsewell** [comment]
thanks @DirectriX01, note your link in the comment above is pointing to the wrong thing

**Review by ltalirz — COMMENTED** [review]
thanks for the PR @DirectriX01 !

The link under "importlib documentation" currently links to the AiiDA source code, I guess you meant to insert a different link here?
A reference on how to adequately replace the `load_package` and `load_source` functions would indeed be useful, perhaps also for the commit message.
I guess you've looked at the [imp source code](https://github.com/python/cpython/blob/1744c96ebc98b240f2564f75191097704b37244f/Lib/imp.py#L200-L217) and taken what seemed necessary?

Anyhow, the tests seem to pass...

**Comment by ltalirz** [comment]
> Anyhow, the tests seem to pass...

Hm... the failing django test does look like it might have to do with this change.
If there is really no easy replacement for these functions, should we vendor them?

**Comment by DirectriX01** [comment]
I'm really sorry for overlooking and copying the wrong link. Here is the [Importlib documentation](https://docs.python.org/3/library/importlib.html#checking-if-a-module-can-be-imported)

File: aiida/restapi/translator/nodes/node.py
Comment (ltalirz) [inline aiida/restapi/translator/nodes/node.py:355]: typo?
```suggestion
                spec = importlib.util.spec_from_file_location(f'{name}', full_path)
```

File: aiida/restapi/translator/nodes/node.py
Comment (ltalirz) [inline aiida/restapi/translator/nodes/node.py:357]: same here (or perhaps I don't understand what this does).

is this some weird autocomplete of "firstname" in the editor?

**Comment by ltalirz** [comment]
By the way, I believe the tests should also run on your fork after you push, i.e. this is an easy way to check whether all tests pass before making a pull request (no worries about this one).

File: aiida/restapi/translator/nodes/node.py
Comment (DirectriX01) [reply aiida/restapi/translator/nodes/node.py]: Yes, and I have changed them accordingly

**Comment by DirectriX01** [comment]
The previous errors were because it couldn't load a package from the `__init__` file. Since there was no direct way in `importlib` module, I had to implement [imp.load_package()](https://github.com/python/cpython/blob/1744c96ebc98b240f2564f75191097704b37244f/Lib/imp.py#L200-L217) from the source code

File: aiida/restapi/translator/nodes/node.py
Comment (ltalirz) [inline aiida/restapi/translator/nodes/node.py:347]:
```suggestion
            # N.B. pkgutil.walk_packages requires a LIST of paths
```

File: aiida/restapi/translator/nodes/node.py
Comment (ltalirz) [inline aiida/restapi/translator/nodes/node.py:352]:
```suggestion
                # re-implementation of deprecated `imp.load_package`
                if os.path.isdir(full_path_base):
                    #Adds an extension to check for __init__ file in the package directory
```

File: aiida/restapi/translator/nodes/node.py
Comment (ltalirz) [inline aiida/restapi/translator/nodes/node.py:372]:
```suggestion
                # reimplementation of deprecated `imp.load_source`
                spec = importlib.util.spec_from_file_location(name, full_path)
                app_module = importlib.util.module_from_spec(spec)
```

**Review by ltalirz — CHANGES_REQUESTED** [review]
thanks for fixing @DirectriX01  , just some minor comments then good to go from my side
@chrisjsewell you agree?

I'm surprised that importlib provides no easier non-deprecated path for this - this is going to cause headaches for a lot of python packages when the functionality is removed...

This made the issue quite a bit more complicated than expected, so in retrospect it was not really a "good first issue" - apologies.

**Comment by ltalirz** [comment]
Hey Abhinav, thanks for taking on this issue and bringing the PR to completion (I consider the work done; just waiting for approval from @chrisjsewell ).

I was wondering: are you still planning to submit a GSOC proposal? As you may be aware, the deadline is in ~3 hours, i.e. it's still possible but a bit tight ;-) 
Feel free to drop me a line via email (you'll find it via my website linked on my github profile).

**Review by chrisjsewell — APPROVED** [review]
cheers!

**Pull request merged**
